### ISSUE TITLE
Output frames from openh264 in chronological order by DecodeFrameNoDelay

### ISSUE THREAD

**Issue by WeiChungChang** [cisco/openh264#3601]
Openh264 does NOT always output video frames in chronological order when calling by DecodeFrameNoDelay()

The output PTSs are as below; some of them are out-of-order. Should I reorder myself?

```
0	0
1	33333
2	66666
3	100000
4	133333
5	166666
6	200000
7	266666 *
8	233333 #
9	300000
10	333333
11	400000 *
12	366666 #
13	433333
14	466666
15	533333 *
16	500000 #
17	566666
18	600000
19	666666 * 
20	633333 #
21	700000
22	733333
23	800000 *
24	766666 #
25	833333
26	900000 *
27	866666 #
28	933333
```

**Comment by huili2** [comment]
Openh264 decoder supports real-time decoding, so the output order might be the same as decoding order. @xiaotiansf Is that also true for B frame decoding?

**Comment by xiaotiansf** [comment]
That's not correct. If there is B-frame, the output order (which is called presentation order) will be different from decoding order. For example, the decoding order is I B B P, the output order has to be B B I ...

**Comment by WeiChungChang** [comment]
@huili2  @xiaotiansf  Thanks to the info.

So the conclusion is when I use DecodeFrameNoDelay() to decode movie with B frames, I need to implement a logic to reoder the decoded frames in chronological order?


If so, I had the simple implementation below(in pseudo code):

```
std::priority_queue<decoded_buffer> pq;
...
SBufferInfo buffer_info;
decoder_->DecodeFrameNoDelay(..., buffer_info);

if (buffer_info.iBufferStatus == 1) {
  Receive current_frame;
  if (!pq.empty() && pq.top().timestamp < current_frame's timestamp) {
    output pq.top();
  }
  pq.push(current_frame);
}
```

I skip the logic in flush.

In this logic, when receive a new decoded frame it will not output it immediately. 
Instead, it inspect if
1. the priority queue is not.
2. the first(in priority queue, the buffer with smallest PTS(oldest)) is smaller than current frame's PTS.

If both 1 & 2 are ture, output the first item of priority queue.

For example, to 

```
6	200000 (1st) -> pq = [200000]
7	266666 (2nd) -> output 200000, push 266666, pq = [266666]
8	233333 (3rd) -> no output, push 233333, pq = [233333, 2666666]
9	300000 (4th) -> output 2333333, pq = [2666666, 300000].
....
```

It output 200000, 233333, 266666, 300000... in  chronological order.

**[Q1]: could you please help to comment if this method covers openh264 DecodeFrameNoDelay()  B frame handling logic and can reorder the decoded frames in chronological order?**

**[Q2]: if so, if it is possible (and suitable) to integrate the logic to help deocder always outputs frames in chronological order as most of the modern decoders?**

Thanks

**Comment by xiaotiansf** [comment]
A1: Your logic is correct as long as you know "current_frame's timestamp", pts.
A2. openh264 has already the mechanism to reorder frames in presentation order in welsDecoderExt.cpp (check it). The logic to reorder is not based on pts since decoder does not have pts available .

**Comment by xiaotiansf** [comment]
What's your video profileidc?

**Comment by xiaotiansf** [comment]
Can you share the video and I can take a look?

**Comment by xiaotiansf** [comment]
Yes. Are you sure that you send the pictures in decoding order to the decoder. I am seeing the following pictures in decoder order:
fprobe ~/Downloads/videoplayback.mp4 -select_streams v -show_entries frame=coded_picture_number,pkt_dts_time,pict_type -of csv=p=0:nk=1 -v 0 | more
0.000000,I,0
0.033333,B,2
0.066667,B,3
0.100000,P,1
0.133333,B,5
0.166667,B,6
0.200000,B,7
0.233333,P,4
0.266667,B,9
0.300000,B,10
0.333333,B,11
0.366667,P,8

**Comment by xiaotiansf** [comment]
Yes. I have noticed that video output from the decoder is not pts order somehow. So it is a bug in openh264.

**Comment by xiaotiansf** [comment]
I tested with if you use threaded decoding by setting threadcount=2 or 3, the output video order is correct. But non-threaded decoding gets incorrect output video order. I will fix the issue for non-threaded decoding.

**Comment by WeiChungChang** [comment]
Hi @xiaotiansf 

The issue starts from this 2 commits, for you reference, thank you!

https://github.com/cisco/openh264/commit/c96de068fcd435579eeb3c63c132d8b8d182964c

https://github.com/cisco/openh264/commit/0819f288dc5fc24fb34089113bc230aea65c8375

**Comment by xiaotiansf** [comment]
Thanks for the checking.
